api/types.go

```diff
@@ -115,3 +115,108 @@ type TaggedMetric struct {
+
+// Timerange represents a range of time a given time series is defined in:
+// it is 3-tuple of (start, end, resolution) with the following constraints:
+// start <= end
+// start = 0 mod resolution
+// end =   0 mod resolution
+type Timerange struct {
+	Start      int64
+	End        int64
+	Resolution int64
+}
+
+// IsValid determines whether the given timerange meets the constraint.
+func (tr Timerange) IsValid() bool {
+	return (tr.Start%tr.Resolution == 0 &&
+		tr.End%tr.Resolution == 0 &&
+		tr.Resolution > 0 &&
+		tr.Start <= tr.End)
+}
+
+// Slots represent the total # of data points
+func (tr Timerange) Slots() int {
+	return int((tr.Start - tr.End) / tr.Resolution)
+}
+
+// Timeseries is a single time series, identified with the associated tagset.
+type Timeseries struct {
+	Values []float64
+	TagSet TagSet
+}
+
+// SamplingStrategy determines how the given time series should be sampled.
+// Note(This is currently unused).
+type SamplingStrategy int
+
+const (
+	// SamplingMax chooses the maximum value.
+	SamplingMax SamplingStrategy = iota + 1
+	// SamplingMin chooses the minimum value.
+	SamplingMin
+	// SamplingMean chooses the average value.
+	SamplingMean
+)
+
+// SeriesResult is the abstract interface type describing the result of a time series operation.
+type SeriesResult interface {
+	// Sample the given result to the given timerange, using the provided sampling strategy.
+	Sample(timerange Timerange, sampling SamplingStrategy) SeriesList
+}
+
+// SeriesList is a list of time series sharing the same time range.
+type SeriesList struct {
+	List      []Timeseries
+	Timerange Timerange
+}
+
+// IsValid determines whether the given time series is valid.
+func (list SeriesList) IsValid() bool {
+	if !list.Timerange.IsValid() {
+		// timerange must be valid.
+		return false
+	}
+	for _, series := range list.List {
+		// # of slots per series must be valid.
+		if len(series.Values) != list.Timerange.Slots() {
+			return false
+		}
+	}
+	return true // validation is now successful.
+}
+
+// Sample converts the given serieslist to comform with the provided sampling strategy.
+func (s SeriesList) Sample(timerange Timerange, sampling SamplingStrategy) SeriesList {
+	// TODO - deal with the different range.
+	return s
+}
+
+// ScalarSeriesList represents a scalar value returned by the query.
+// such scalar value can be converted to a SeriesList of any timerange.
+type ScalarSeriesList float64
+
+// Sample converts a given scalar to a SeriesList with a single series, repeating the same value.
+func (s ScalarSeriesList) Sample(timerange Timerange, sampling SamplingStrategy) SeriesList {
+	slots := timerange.Slots()
+	value := float64(s)
+	values := make([]float64, slots)
+	for i := 0; i < slots; i++ {
+		values[i] = value
+	}
+	return SeriesList{
+		List:      []Timeseries{Timeseries{values, NewTagSet()}},
+		Timerange: timerange,
+	}
+}
+
+// MetricMetadata is metadata associated with the given metric.
+type MetricMetadata struct {
```

Comment:
Would this be more useful as { seriesName : { meta_key : meta_value } }? Or otherwise passing this information closer to what it's supposed to be metadata for? Parallel arrays is a bit weird.